### PR TITLE
[DONOTMERGE] Profile resolver error handling

### PR DESCRIPTION
**Pull request by galtm**
# Committer Notes

This PR enhances the XSLT profile resolver so that (a) if there is a fatal error, the resolver reports one such error and does not produce a final output, and (b) if there are no fatal errors, the resolver reports all nonfatal error and warning messages when producing the final output document.

1. Separate before/after from starting/ending in modify phase because they are no longer synonyms
1. Enhance message-handler.xsl for pipeline usage and testability
1. Adapt XSLT to message handler returning PI
1. Enable saving intermediate XML files for debugging purposes
1. Add sample document for nonfatal errors/warnings

Included bug fixes for bugs I noticed during this work:
* Enable param to have insertions before/after AND have param set
* Fix merge phase bug where descendant controls weren't being flattened

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/usnistgov/OSCAL/blob/main/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers
](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/OSCAL/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- ~Have you included examples of how to use your new feature(s)?~
- ~Have you updated all [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes you made? Changes to the OSCAL website can be made in the docs/content directory of your branch.~

## Testing Done

While qualifying these changes, I ran commands like the following in a Windows shell, from the `resolver-pipeline` directory.
```
java -cp C:\SaxonHE10-8J\saxon-he-10.8.jar net.sf.saxon.Transform -t -s:..\..\..\specifications\profile-resolution\profile-resolution-examples\broken_profile.xml -xsl:oscal-profile-RESOLVE.xsl -o:myoutput.xml
```
The command window output shows the error message as in the following example.
```
Terminating Error: Document not acquired: missing_profile.xml resolved as file:/C:/github/oscal/OSCAL/src/specifications/profile-resolution/profile-resolution-examples/missing_profile.xml (as OSCAL XML)
Error at xsl:message on line 150 column 60 of oscal-profile-RESOLVE.xsl:
  XTMM9000  Processing terminated by xsl:message at line 150 in oscal-profile-RESOLVE.xsl
     invoked by xsl:for-each at file:/C:/github/oscal/OSCAL/src/utils/util/resolver-pipeline/oscal-profile-RESOLVE.xsl#149
  In template rule with match="opr:terminate-if-severe-errors" on line 144 of oscal-profile-RESOLVE.xsl
     invoked by xsl:apply-templates at file:/C:/github/oscal/OSCAL/src/utils/util/resolver-pipeline/oscal-profile-RESOLVE.xsl#81
     invoked by xsl:iterate at file:/C:/github/oscal/OSCAL/src/utils/util/resolver-pipeline/oscal-profile-RESOLVE.xsl#77
  In template rule with match="/" on line 69 of oscal-profile-RESOLVE.xsl
Processing terminated by xsl:message at line 150 in oscal-profile-RESOLVE.xsl
```

In another command-line example with nonfatal errors and warnings, the command window output includes the messages near the end:
```
Warning: Found 2 elements of type &lt;title&gt;. Only the last will be present in result.
Error: Element of type &lt;back-matter&gt; is not valid inside &lt;control&gt; and will not be present in result.
Warning: set-parameter with param-id="a3_prm1" does not match any param id.
Execution time: 644.6529ms
Memory used: 30Mb
```

Behavior when using Oxygen transformation scenario: Messages appear in the Messages pane. (In the case of fatal error messages, this pane is underneath the Transformation problems pane.)

### CODE REVIEW

File: src/utils/util/resolver-pipeline/oscal-profile-resolve-finish.xsl
Comment (galtm) [inline src/utils/util/resolver-pipeline/oscal-profile-resolve-finish.xsl:63]: Most changes in this file are aimed at reporting when an element doesn't belong instead of silently dropping it.

File: src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl
Comment (galtm) [inline src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl:50]: Bug fix. Earlier behavior undid the grouping but left the nesting.

File: src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl
Comment (galtm) [inline src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl:79]: Example of the kind of data type change required by Item 3 in PR description.

File: src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl
Comment (galtm) [inline src/utils/util/resolver-pipeline/oscal-profile-resolve-merge.xsl:93]: Item 3 in PR description also required some handling of the processing instructions so they didn't get lost partway through the pipeline.

File: src/utils/util/resolver-pipeline/oscal-profile-resolve-modify.xsl
Comment (galtm) [inline src/utils/util/resolver-pipeline/oscal-profile-resolve-modify.xsl:55]: Part of the bug fix to enable param to have insertions before/after AND have param set

File: src/specifications/profile-resolution/profile-resolution-examples/nonfatal-error-warning_profile.xml
Comment (galtm) [inline src/specifications/profile-resolution/profile-resolution-examples/nonfatal-error-warning_profile.xml:1]: Item 5 in PR description

**Review by galtm — COMMENTED** [review]
Some annotations to make large diffs more manageable

**Comment by galtm** [comment]
@david-waltermire-nist , @wendellpiez : I'm a little concerned about potential conflicts between this PR and #1424 (which I haven't looked at in detail, but I see some of the same files). If I can help make file merges go more smoothly, let me know.

**Review by wendellpiez — APPROVED** [review]
Thanks again for all the tremendous work here, it's looking nice!

**Comment by aj-stein-nist** [comment]
Hi @galtm, thanks for submitting yet another excellent PR. We had to rebase develop and that leads to the "This branch has conflicts that must be resolved" conflicts below. I can help you with that, please let me know if and when you need that help. Thanks.

**Comment by galtm** [comment]
Hi, @aj-stein-nist . Thanks for notifying me and offering to help. The files listed are files that I haven't touched, which should make it easier. I'll try to look into this today or tomorrow and I'll keep you posted.

**Comment by galtm** [comment]
@aj-stein-nist : I resolved the merge conflicts. If anything doesn't look right to you, let me know. Thanks.

@wendellpiez : When I did the force-push, GitHub "dismissed" your earlier code review.

**Comment by david-waltermire-nist** [comment]
> @david-waltermire-nist , @wendellpiez : I'm a little concerned about potential conflicts between this PR and #1424 (which I haven't looked at in detail, but I see some of the same files). If I can help make file merges go more smoothly, let me know.

We should figure out how to move forward with this given your concerns.

Would it make sense to move this PR against #1424?

**Comment by galtm** [comment]
> Would it make sense to move this PR against #1424?

Would that mean that this PR can't be merged until #1424 is merged? #1424 is a draft and has been open for about 5 months. To keep things moving, can you merge this change onto develop and rebase #1424 on top?

If I could understand the essence of what is changing with #1424 I could look into whether there are any potential merge conflicts with this PR that might be hard to resolve. However, I don't know how to make sense of the 56 commits that probably resulted from rebasing the develop branch. Which commits should I be looking at in that list?

**Comment by david-waltermire-nist** [comment]
> > Would it make sense to move this PR against #1424?
> 
> Would that mean that this PR can't be merged until #1424 is merged? #1424 is a draft and has been open for about 5 months. To keep things moving, can you merge this change onto develop and rebase #1424 on top?
> 
> If I could understand the essence of what is changing with #1424 I could look into whether there are any potential merge conflicts with this PR that might be hard to resolve. However, I don't know how to make sense of the 56 commits that probably resulted from rebasing the develop branch. Which commits should I be looking at in that list?

I think your suggestion of merging this now and rebasing #1424 on top is a good path.

**Review by david-waltermire-nist — DISMISSED** [review]

**Review by wendellpiez — DISMISSED** [review]
Beautiful work here, and much interesting progress.

**Comment by galtm** [comment]
2/4/23: Pushed 2 commits to fix minor issues I noticed while working on something unrelated.

**Comment by galtm** [comment]
3/9/23: Manually resolved merge conflicts now that #1685 has been merged into develop branch.

File: src/utils/util/resolver-pipeline/testing/4_modified/modify.xspec
Comment (galtm) [inline src/utils/util/resolver-pipeline/testing/4_modified/modify.xspec:1608]: @wendellpiez : Does this test scenario cover the situation in issue #1859?

Comment (wendellpiez) [reply src/utils/util/resolver-pipeline/testing/4_modified/modify.xspec:1608]: I don't think so ... what is being described is wanting to add a new `prop` to an existing `param`, but not being able to use `alter`. Apparently `alter` is a no-op against params. `set-param` is the available workaround (it can be used to add new content). But the spec does not say that you can't target `param` elements using `alter/add`, so that should also work.

Comment (galtm) [reply src/utils/util/resolver-pipeline/testing/4_modified/modify.xspec:1608]: Thanks. You're right, this scenario isn't the same as what the issue is about. But I think the XSLT in this PR fixes the issue. If I create a similar test scenario with
```
<alter control-id="a1">
    <add position="starting" by-id="a1_prm1">
        <prop ns="https://fedramp.gov/ns/oscal" name="param-label" value="AC-1(a)"/>
    </add>
</alter>
```
then I can see that the new `prop` gets added at the beginning of the existing `param`, like this:
```
<param id="a1_prm1">
    <prop ns="https://fedramp.gov/ns/oscal" name="param-label" value="AC-1(a)"/>
    <label>A1 Parameter 1</label>
    <constraint>at least every 3 years</constraint>
</param>
```
Should I add the new test scenario?

Comment (wendellpiez) [reply src/utils/util/resolver-pipeline/testing/4_modified/modify.xspec:1608]: Sure! I guess I was not looking closely enough, or not looking at the current code, to see that this should actually work.

If it actually works, so much the better, we only need to be able to show that it does.

Comment (galtm) [reply src/utils/util/resolver-pipeline/testing/4_modified/modify.xspec]: I pushed commit 5d3b9b302d2d06080afeee5eb7a6aa6d176cfaf2 to this PR branch, to provide a test of the situation in #1859. I also tweaked the label of the other test scenario to distinguish the two scenarios from each other more.

**Comment by aj-stein-nist** [comment]
Well, this is frustrating, I have a backup of this branch before rebase, but I am unable to push to your fork @galtm after "aligning" it to develop (with `--strategy=ours` to save time) but that did not work.

@galtm, can you check out my branch and push it here to reopen? Shorthand instructions below. Thanks. I was trying to spare you the effort of rebasing this yourself.

```sh
# Just for alignment on what remotes I mean
git add remote me git@github.com:galtm/OSCAL.git
git add remote aj git@github.com:aj-stein-nist/OSCAL.git
git fetch all
git checkout --track aj/profiler-error-handling-backup
git push me profiler-error-handling-backup:profiler-error-handling -f
```

Then you should be able to continue on this PR by ID and receive the credit you deserve. Apologies.

**Comment by aj-stein-nist** [comment]
Reopening per the comment above.

**Comment by galtm** [comment]
@aj-stein-nist : I tried to follow your shorthand instructions, but I'm not sure if things worked. Here's the output I got from the push:
```
$ git push origin profiler-error-handling-backup:profiler-error-handling -f
Enumerating objects: 192, done.
Counting objects: 100% (157/157), done.
Delta compression using up to 8 threads
Compressing objects: 100% (81/81), done.
Writing objects: 100% (112/112), 53.52 KiB | 3.57 MiB/s, done.
Total 112 (delta 84), reused 44 (delta 26), pack-reused 0
remote: Resolving deltas: 100% (84/84), completed with 33 local objects.
To https://github.com/galtm/OSCAL.git
 + 0825b9fc4...5d3b9b302 profiler-error-handling-backup -> profiler-error-handling (forced update)
```
where origin is as follows (the `git@github.com:...` syntax didn't work for me when I tried to fetch).
```
origin  https://github.com/galtm/OSCAL.git (fetch)
origin  https://github.com/galtm/OSCAL.git (push)
```

Despite the lack of error messages, this PR page still says 0 commits. I thought it would add commits to this PR but they'd be aligned with where `develop` stands now. Did I misunderstand the result of those commands or what I'm supposed to do next?

It may be simpler for me to sync my develop branch, create a fresh branch from that, do a file-merge (outside Git) of the profile resolver file contents based on the files in a backup branch, and try again to push to the branch of this PR. I started doing that but still need to check the results. I'll be busy for a few hours, but I will come back to this in the late afternoon.

**Comment by aj-stein-nist** [comment]
It seems now I can reopen?

**Comment by aj-stein-nist** [comment]
Well ti seems even though I have reopened it and commented it seems to ignore the state of the branch on your end even though it looks just as before (155 commits off of develop and the same series of changes I worked through last night). I guess I will ask to reopen the branch.

**Comment by aj-stein-nist** [comment]
Apologize, I guess we move #1943 in favor of this PR, I cannot even reopen it. Apologies and sincere thanks, @galtm.

**Comment by galtm** [comment]
I'm confused, too, because I thought rebasing against develop would mean that the list of commits and changed files would look like my work in isolation rather than the long list that I see in #1943.

I also realized that the code needs a few tweaks because the profiler resolver code is less deep in the directory hierarchy than it used to be.

I created #1944 with what I think is the right code, on top of the develop branch. Does it look right from your perspective?

**Comment by aj-stein-nist** [comment]
> I'm confused, too, because I thought rebasing against develop would mean that the list of commits and changed files would look like my work in isolation rather than the long list that I see in #1943.
> 
> I also realized that the code needs a few tweaks because the profiler resolver code is less deep in the directory hierarchy than it used to be.
> 
> I created #1944 with what I think is the right code, on top of the develop branch. Does it look right from your perspective?

I am looking at in a few minutes, we just wrapped our sprint review, thanks again @galtm.

**Pull request closed without merging**
